jujugui/static/gui/src/app/app.js

```diff
@@ -669,6 +662,112 @@ YUI.add('juju-gui', function(Y) {
+      the Juju controller if it's necessary. A username and password must be
+      supplied if you're connecting to a standalone Juju controller and not to
+      one which requires macaroon authentication.
+
+      @method setUpControllerAPI
+      @param {Object} controllerAPI Instance of the GoEnvironment class.
+      @param {String} user The username if not using macaroons.
+      @param {String} password The password if not using macaroons.
+      @param {Array} macaroons A list of macaroons that the user has saved.
+      @return {environments.GoEnvironment} An instance of the environment class
+        with the necessary events attached and values set. Or undefined if
+        we're in a legacy juju model and no controllerAPI instance was supplied.
+    */
+    setUpControllerAPI: function(controllerAPI, user, password, macaroons) {
+      // If no instance was supplied to us then we're in a legacy Juju
+      // and this connection is not necessary.
+      if (!controllerAPI) { return; }
+
+      controllerAPI.setAttrs({ user, password });
+      controllerAPI.setCredentials({ user, password, macaroons });
+
+      controllerAPI.after('login', e => {
+        this.controllerAPI.listModelsWithInfo((err, response) => {
+          if (err) {
+            console.error('Unable to list models', err);
+            this.db.notifications.add({
+              title: 'Unable to list models',
+              message: 'Unable to list models: ' + err,
+              level: 'error'
+            });
+            return;
+          }
+          const modelList = response.models;
+          this.set('environmentList', modelList);
+          // If the modelList has no models in it then we have to drop the
+          // user into an uncommitted state.
+          if (modelList.length === 0) {
+            // XXX Drop the user into the uncommitted state.
+          }
+          const selectedModel = this._pickModel(modelList);
+          this.switchEnv(this.createSocketURL(selectedModel.uuid));
+        });
+      });
+
+      controllerAPI.after('connectedChange', e => {
+        const credentials = this.controllerAPI.getCredentials();
+        if (!credentials.areAvailable) {
+          // If we don't have credentials then do nothing as the env login
+          // check will have kicked the user to the login prompt already and
+          // we can wait until they have provided the credentials there.
+          return;
+        }
+        // If we're in a JIMM controlled environment or if we have macaroon
+        // credentials then use the macaroon login. If not then uses the
+        // standard u/p method.
+        if (this.get('jimmURL') || credentials.macaroons) {
+          this.loginToAPIs(null, true, [this.controllerAPI]);
+        } else {
+          this.loginToAPIs(null, false, [this.controllerAPI]);
+        }
+      });
+      // If we're in JIMM then use a jimmURL, else use a socket_url without the
```

Comment:
Also, I think we are not currently using the controller socket template for generating the URL where to connect the controller. Is this a follow-up card? So please add a TODO and a card.